### ISSUE TITLE
WAL information may not be added to the xlog.db file

### ISSUE THREAD

**Issue by albatross0** [EnterpriseDB/barman#328]
barman v2.11 (RPM-installed on CentOS 8)

The following lines are part of my xlog.db file.

```
000000020002E59C000000BC        3089819 1615159937.4639714      pybzip2
000000020002E59C000000BD        3071435 1615159938.1930323      pybzip2
000000020002E59C000000BE        3092476 1615159938.944095       pybzip2
000000020002E59C000000C0        3136510 1615159940.656238       pybzip2
000000020002E59C000000C1        3286609 1615159944.1185274      pybzip2
000000020002E59C000000C2        3201442 1615159944.4345536      pybzip2
```

The line of 000000020002E59C000000BF is missing, but actually the archived WAL file exists.

```
# cd /var/lib/barman/dbname/wals/000000020002E59C/
# ls -l 000000020002E59C000000BF
-rw-------. 1 barman barman 3148731  3月  8 08:32 000000020002E59C000000BF
```

**Comment by albatross0** [comment]
To investigate, I added logging code to xlogdb function as below.

```diff
--- server.py.orig       2020-07-08 18:12:30.000000000 +0900
+++ server.py   2021-02-03 10:51:00.582264786 +0900
@@ -2306,6 +2306,7 @@
                     mode = "a%s" % mode[1:]

             with open(xlogdb, mode) as f:
+                _logger.info("%s stat=%s", xlogdb, os.stat(f.fileno()))

                 # execute the block nested in the with statement
                 try:
```

And I got the following log.

```
2021-03-08 08:35:12,883 [4011354] barman.wal_archiver INFO: Found 362 xlog segments from file archival for dbname. Archive all segments in one run.

...

2021-03-08 08:49:02,643 [4015392] barman.server INFO: Another archive-wal process is already running on server dbname. Skipping to the next server
2021-03-08 08:49:02,673 [4015392] barman.retention_policies INFO: Reporting backup 20210205T050102 for server dbname as OBSOLETE (older than 2021-02-06 08:49:02.672772+09:00)
2021-03-08 08:49:02,674 [4015392] barman.backup INFO: Enforcing retention policy: removing backup 20210205T050102 for server dbname
2021-03-08 08:49:02,680 [4015392] barman.backup INFO: Deleting backup 20210205T050102 for server dbname
2021-03-08 08:49:04,492 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863398, st_atime=1615160942, st_mtime=1615160941, st_ctime=1615160941)
2021-03-08 08:49:04,577 [4011354] barman.wal_archiver INFO: Archiving segment 290 of 362 from file archival: dbname/000000020002E59C000000BB
2021-03-08 08:49:07,516 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863458, st_atime=1615160942, st_mtime=1615160944, st_ctime=1615160944)
2021-03-08 08:49:07,560 [4011354] barman.wal_archiver INFO: Archiving segment 291 of 362 from file archival: dbname/000000020002E59C000000BC
2021-03-08 08:49:10,201 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863518, st_atime=1615160942, st_mtime=1615160947, st_ctime=1615160947)
2021-03-08 08:49:10,296 [4011354] barman.wal_archiver INFO: Archiving segment 292 of 362 from file archival: dbname/000000020002E59C000000BD
2021-03-08 08:49:12,511 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863578, st_atime=1615160942, st_mtime=1615160950, st_ctime=1615160950)
2021-03-08 08:49:12,675 [4011354] barman.wal_archiver INFO: Archiving segment 293 of 362 from file archival: dbname/000000020002E59C000000BE
2021-03-08 08:49:15,569 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863638, st_atime=1615160942, st_mtime=1615160952, st_ctime=1615160952)
2021-03-08 08:49:15,605 [4011354] barman.wal_archiver INFO: Archiving segment 294 of 362 from file archival: dbname/000000020002E59C000000BF
2021-03-08 08:49:15,642 [4015392] barman.backup INFO: Delete associated WAL segments:
2021-03-08 08:49:15,643 [4015392] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863697, st_atime=1615160942, st_mtime=1615160955, st_ctime=1615160955)
2021-03-08 08:49:15,657 [4015392] barman.backup INFO: /var/lib/barman/dbname/wals/xlog.db.new stat=os.stat_result(st_mode=33188, st_ino=7058292408, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=0, st_atime=1615160955, st_mtime=1615160955, st_ctime=1615160955)
2021-03-08 08:49:37,197 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_result(st_mode=33188, st_ino=1613568054, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15863697, st_atime=1615160955, st_mtime=1615160955, st_ctime=1615160955)
2021-03-08 08:49:37,221 [4015392] barman.backup INFO:   000000020002E19000000058
2021-03-08 08:49:37,222 [4015392] barman.backup INFO:   000000020002E19000000059
2021-03-08 08:49:37,223 [4015392] barman.backup INFO:   000000020002E1900000005A

...

2021-03-08 08:49:37,490 [4015392] barman.backup INFO:   000000020002E19100000047
2021-03-08 08:49:37,491 [4015392] barman.backup INFO:   000000020002E19100000048
2021-03-08 08:49:37,492 [4011354] barman.wal_archiver INFO: Archiving segment 295 of 362 from file archival: dbname/000000020002E59C000000C0
2021-03-08 08:49:37,492 [4015392] barman.backup INFO:   000000020002E19100000049
2021-03-08 08:49:37,493 [4015392] barman.backup INFO:   000000020002E1910000004A
2021-03-08 08:49:37,494 [4015392] barman.backup INFO:   000000020002E1910000004B

...

2021-03-08 08:49:41,118 [4015392] barman.backup INFO:   000000020002E19E00000001
2021-03-08 08:49:41,119 [4015392] barman.backup INFO:   000000020002E19E00000002
2021-03-08 08:49:41,120 [4015392] barman.backup INFO:   000000020002E19E00000003
2021-03-08 08:49:41,122 [4011354] barman.server INFO: /var/lib/barman/dbname/wals/xlog.db stat=os.stat_r
esult(st_mode=33188, st_ino=7058292408, st_dev=64768, st_nlink=1, st_uid=992, st_gid=987, st_size=15373313
, st_atime=1615160955, st_mtime=1615160977, st_ctime=1615160977)
2021-03-08 08:49:41,122 [4015392] barman.backup INFO:   000000020002E19E00000004
2021-03-08 08:49:41,123 [4015392] barman.backup INFO:   000000020002E19E00000005
2021-03-08 08:49:41,124 [4015392] barman.backup INFO:   000000020002E19E00000006
```

The log shows the following behavior.
1. The archive-wal process (PID:4011354) was running. (forked from `barman cron`)
2. The backup process (PID:4015392) started deleting WAL segments. (forked from `barman backup`)
   This process opened xlog.db (inode: 1613568054) and xlog.db.new (inode: 7058292408)
3. Then the archive-wal process (PID:4011354) opened xlog.db (inode: 1613568054) and wrote the line of 000000020002E59C000000BF to (old) xlog.db file.

**Comment by amenonsen** [comment]
@albatross0 thanks for the detailed analysis, and the proposed fix in #329. However, after some research, I don't think it is safe to use `shutil.move()` inside the locked section to overwrite xlog.db, so I am going to fix this differently for the 2.13 release.
